src/iceberg/catalog/in_memory_catalog.cc

```diff
@@ -0,0 +1,496 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "iceberg/catalog/in_memory_catalog.h"
+
+#include <algorithm>
+#include <iterator>  // IWYU pragma: keep
+#include <mutex>
+#include <optional>
+#include <unordered_map>
+
+#include "iceberg/exception.h"
+#include "iceberg/table.h"
+#include "iceberg/util/macros.h"
+
+namespace iceberg {
+
+namespace {
+
+/**
+ * \brief A hierarchical namespace that manages namespaces and table metadata in-memory.
+ *
+ * Each InMemoryNamespace represents a namespace level and can contain properties,
+ * tables, and child namespaces. This structure enables a tree-like representation
+ * of nested namespaces.
+ */
+class ICEBERG_EXPORT InMemoryNamespace {
```

Comment:
As discussed in https://github.com/apache/iceberg-cpp/pull/80#discussion_r2059685178, we need to move some of these functions (which are included in the rest catalog spec) as a part of the public Catalog API.